test/tests/functional/pbs_provisioning_enhancement.py

```diff
@@ -0,0 +1,310 @@
+             'resources_available.aoe': 'App1,osimage1'}
+        self.server.manager(
+            MGR_CMD_SET, NODE, a, id=self.hostA, expect=True)
+        self.server.manager(MGR_CMD_UNSET, NODE, id=self.hostA,
+                            attrib='current_aoe', expect=True)
+
+        # Set hostB ncpus to 12
+        a = {'resources_available.ncpus': '12'}
+        self.server.manager(
+            MGR_CMD_SET, NODE, a, id=self.hostB, expect=True)
+
+        self.server.manager(
+            MGR_CMD_SET, SERVER, {
+                'log_events': 2047}, expect=True)
+        self.server.expect(NODE, {'state': 'free'}, id=self.hostA)
+
+        # Setup provisioning hook.
+        a = {'event': 'provision', 'enabled': 'True', 'alarm': '300'}
+        rv = self.server.create_import_hook(
+            "fake_prov", a, self.hooks['fake_prov'], overwrite=True)
+        self.assertTrue(rv)
+
+    def test_app_provisioning(self):
+        """
+        Test application provisioning
+        """
+        j = Job(TEST_USER1)
+        j.set_sleep_time(5)
+        j.set_attributes({'Resource_List.select': '1:aoe=App1'})
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        self.server.log_match(
+            "fake application provisioning script",
+            max_attempts=20,
+            interval=1)
+
+    def test_os_provisioning(self):
+        """
+        Test os provisioning
+        """
+        j = Job(TEST_USER1)
+        j.set_sleep_time(10)
+        j.set_attributes({'Resource_List.select': '1:aoe=osimage1'})
+        jid = self.server.submit(j)
+
+        # Job will start and wait for provisioning to complete.
+        self.server.expect(JOB, {ATTR_substate: '71'}, id=jid)
+        self.server.log_match("fake os provisioning script",
+                              max_attempts=60,
+                              interval=1)
+        self.momA.restart()
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+
+    def test_subchunk_application_provisioning(self):
+        """
+        Test application provisioning job request consist of subchunks
+        with and without aoe resource.
+        """
+
+        j = Job(TEST_USER1)
+        j.set_attributes({'Resource_List.select':
+                          '1:ncpus=1:aoe=App1+1:ncpus=12'})
+        j.set_sleep_time(5)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+        self.server.expect(JOB, ATTR_execvnode, id=jid, op=SET)
+        nodes = j.get_vnodes(j.exec_vnode)
+        self.server.log_match("fake application provisioning script",
+                              max_attempts=20,
+                              interval=1)
+        self.assertTrue((nodes[0] == self.momA.shortname and
+                         nodes[1] == self.momB.shortname) or
+                        (nodes[0] == self.momB.shortname and
+                         nodes[1] == self.momA.shortname))
```

Comment:
If you wanted to, you could force the order in the exec_vnode.  Set a priority on your nodes.

Reply:
Yeah, but i want to make sure the right selection happens through scheduler, instead of me forcing the priority of node eligibility check.

Reply:
It's really the same thing.  Right now the scheduler has the flexibility of doing A+B or B+A.  You are checking for both cases.  I'm just saying you could force it to be A+B and only have to check for one case.  Up to you though.